chain/src/txhashset/txhashset.rs

```diff
@@ -474,6 +488,69 @@ impl TxHashSet {
 		);
 		Ok(())
 	}
+
+	/// Initialize the kernel_pos index.
+	/// Find the block header at the kernel index horizon (14 days of history).
+	/// Then iterate over all kernels since then and add to the index.
+	pub fn init_kernel_pos_index(
+		&self,
+		header_pmmr: &PMMRHandle<BlockHeader>,
+		batch: &Batch<'_>,
+	) -> Result<(), Error> {
+		let now = Instant::now();
+		let head_header = batch.head_header()?;
+		let cutoff_height = head_header
+			.height
+			.saturating_sub(global::kernel_index_horizon().into());
+		let cutoff_hash = header_pmmr.get_header_hash_by_height(cutoff_height)?;
+		let cutoff_header = batch.get_block_header(&cutoff_hash)?;
+		let cutoff_pos = if cutoff_header.height == 0 {
+			1
+		} else {
+			batch.get_previous_header(&cutoff_header)?.kernel_mmr_size + 1
+		};
+		let mut kernel_count = 0;
+		let mut current_header = cutoff_header;
+		for pos in cutoff_pos..=self.kernel_pmmr_h.last_pos {
+			while pos > current_header.kernel_mmr_size {
+				let next_hash = header_pmmr.get_header_hash_by_height(current_header.height + 1)?;
+				current_header = batch.get_block_header(&next_hash)?;
+			}
+			if pmmr::is_leaf(pos) {
+				if let Some(kernel) = self.kernel_pmmr_h.backend.get_data(pos) {
+					batch.save_kernel_pos_height(&kernel.excess, pos, current_header.height)?;
+					kernel_count += 1;
+				}
+			}
+		}
+		debug!(
+			"init_kernel_pos_index: {} kernels, took {}s",
+			kernel_count,
+			now.elapsed().as_secs()
+		);
+		Ok(())
+	}
+
+	/// Compact the kernel_pos index by removing any entries older than the kernel index horizon.
+	/// This will ensure we always have 14 days of kernel history in the index.
+	fn compact_kernel_pos_index(&self, batch: &Batch<'_>) -> Result<(), Error> {
+		let now = Instant::now();
+		let head_header = batch.head_header()?;
+		let cutoff_height = head_header
+			.height
+			.saturating_sub(global::kernel_index_horizon().into());
+		let deleted = batch
+			.kernel_pos_iter()?
+			.filter(|(_, (_, height))| *height < cutoff_height)
+			.map(|(key, _)| batch.delete(&key))
+			.count();
```

Comment:
Out of curiosity I'm wondering if there is a more efficient way to do that. Like some kind of `filter_map`.
 Also we are loosing the error of the batch delete without any logging?

Reply:
What do you mean in terms of "efficient"? 
We need to iterate over the full set of entries in the db.